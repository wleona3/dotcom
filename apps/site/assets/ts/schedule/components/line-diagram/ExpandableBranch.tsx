import { max, times } from "lodash";
import React, { ReactElement, useState, useContext } from "react";
import ExpandableBlock from "../../../components/ExpandableBlock";
import { LineDiagramStop } from "../__schedule";
import { CommonLineDiagramProps } from "./__line-diagram";
import { StopRefContext } from "./graphics/useStopPositions";
import { diagramWidth } from "./line-diagram-helpers";
import {
  CIRC_DIAMETER,
  CIRC_RADIUS,
  BRANCH_LINE_WIDTH
} from "./graphics/graphic-helpers";
import { routeToModeName } from "../../../helpers/css";
import StopCard from "./StopCard";

const BranchToggle = (
  stops: LineDiagramStop[],
  isExpanded: boolean
): JSX.Element => {
  const maxBranches = max(stops.map(ld => ld.stop_data.length - 1)) || 1;
  const width = diagramWidth(maxBranches) + 2;
  return (
    <div className="m-schedule-diagram__expander-header">
      <div style={{ width }}>
        {!isExpanded && (
          <svg
            className={`line-diagram-svg__toggle ${routeToModeName(
              stops[0].route_stop.route!
            )}`}
            width={width}
            height="100%"
          >
            <g
              transform={`translate(${maxBranches * BRANCH_LINE_WIDTH - 1}, 1)`}
            >
              <rect width={CIRC_DIAMETER * 2} height="42" rx={CIRC_DIAMETER} />
              {times(3, i => (
                <circle
                  key={i}
                  cx={CIRC_DIAMETER}
                  cy={10 * (i + 1) + 1}
                  r={CIRC_RADIUS}
                />
              ))}
            </g>
          </svg>
        )}
      </div>
      <button className="btn btn-link m-schedule-diagram__toggle" type="button">
        {stops.length} stops
      </button>
    </div>
  );
};

const ExpandableBranch = (
  props: CommonLineDiagramProps
): ReactElement<HTMLElement> => {
  const { stops, handleStopClick, liveData } = props;
  const [isExpanded, setIsExpanded] = useState(false);
  const updateAllStopCoords = useContext(StopRefContext)[1];

  return (
    <div className="m-schedule-diagram__expander">
      <ExpandableBlock
        header={{
          iconSvgText: null,
          text: BranchToggle(stops, isExpanded)
        }}
        initiallyExpanded={isExpanded}
        notifyExpanded={(blockIsExpanded: boolean) => {
          setIsExpanded(blockIsExpanded);
          updateAllStopCoords();
        }}
        preventScroll
        id={`${stops[0].route_stop.branch}`}
      >
        <>
          {stops.map(stop => (
            <StopCard
              key={stop.route_stop.id}
              stop={stop}
              onClick={handleStopClick}
              liveData={liveData?.[stop.route_stop.id]}
            />
          ))}
        </>
      </ExpandableBlock>
    </div>
  );
};

export default ExpandableBranch;
