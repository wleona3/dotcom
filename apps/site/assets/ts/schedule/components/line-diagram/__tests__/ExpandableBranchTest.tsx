import React from "react";
import { cloneDeep, merge } from "lodash";
import { RouteType } from "../../../../__v3api";
import { LineDiagramStop } from "../../__schedule";
import simpleLineDiagram from "./lineDiagramData/simple.json"; // not a full line diagram
import outwardLineDiagram from "./lineDiagramData/outward.json"; // not a full line diagram
import ExpandableBranch from "../ExpandableBranch";
import { mockedRenderWithStopPositionContext } from "../../../../__tests__/test-helpers";
import { fireEvent, screen } from "@testing-library/dom";

const lineDiagram = (simpleLineDiagram as unknown) as LineDiagramStop[];
let lineDiagramBranchingOut = (outwardLineDiagram as unknown) as LineDiagramStop[];

const route = {
  type: 3 as RouteType,
  name: "route 1",
  long_name: "route 1 long name",
  color: "F00B42",
  id: "route-1",
  direction_names: {
    0: "Outbound",
    1: "Inbound"
  },
  direction_destinations: {
    0: "Begin",
    1: "End"
  },
  description: "key_bus_route",
  "custom_route?": false,
  header: "",
  alerts: []
};

lineDiagram.forEach(({ route_stop }) => {
  route_stop.route = cloneDeep(route);
});

lineDiagramBranchingOut.forEach(({ route_stop }) => {
  route_stop.route = cloneDeep(route);
});

let lineDiagramBranchingIn = cloneDeep(lineDiagramBranchingOut).reverse();
const CRroute = merge(cloneDeep(route), { type: 2 as RouteType });
lineDiagramBranchingIn.forEach(({ route_stop }) => {
  route_stop.route = CRroute;
  if (route_stop["is_terminus?"]) {
    route_stop["is_beginning?"] = !route_stop["is_beginning?"];
  }
});

const handleStopClick = () => {};
const liveData = {};
describe("ExpandableBranch", () => {
  let asFragment: () => DocumentFragment, container: HTMLElement;
  beforeEach(() => {
    ({ asFragment, container } = mockedRenderWithStopPositionContext(
      <ExpandableBranch
        stops={lineDiagram}
        handleStopClick={handleStopClick}
        liveData={liveData}
      />,
      lineDiagram
    ));
  });

  it("renders and matches snapshot", () => {
    expect(asFragment()).toMatchSnapshot();
  });

  it("expands a branch", async () => {
    let moreStops = container.querySelector(
      ".c-expandable-block__panel .m-schedule-diagram__stop"
    );
    expect(moreStops).toBeFalsy();
    fireEvent.click(await screen.findByRole("button"));
    expect(moreStops).toBeTruthy();
  });
});
