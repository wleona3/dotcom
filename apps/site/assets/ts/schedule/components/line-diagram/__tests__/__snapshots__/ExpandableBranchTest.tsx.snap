// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpandableBranch renders and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="m-schedule-diagram__expander"
  >
    <h3
      aria-controls="panel-null"
      aria-expanded="false"
      class="c-expandable-block__header"
      id="header-null"
      role="button"
      tabindex="0"
    >
      <div
        class="c-expandable-block__header-text"
      >
        <div
          class="m-schedule-diagram__expander-header"
        >
          <div
            style="width: 38px;"
          >
            <svg
              class="line-diagram-svg__toggle bus"
              height="100%"
              width="38"
            >
              <g
                transform="translate(15, 1)"
              >
                <rect
                  height="42"
                  rx="8"
                  width="16"
                />
                <circle
                  cx="8"
                  cy="11"
                  r="4"
                />
                <circle
                  cx="8"
                  cy="21"
                  r="4"
                />
                <circle
                  cx="8"
                  cy="31"
                  r="4"
                />
              </g>
            </svg>
          </div>
          <button
            class="btn btn-link m-schedule-diagram__toggle"
            type="button"
          >
            4 stops
          </button>
        </div>
        <span
          class="fa c-expandable-block__header-caret"
        >
          ï„‡
        </span>
      </div>
    </h3>
    <noscript />
  </div>
</DocumentFragment>
`;
